core/bridge/helidon/src/main/kotlin/org/http4k/bridge/HelidonToHttp4kWebSocketListener.kt

```diff
@@ -1,71 +1,61 @@
+    private lateinit var websocket: PushPullAdaptingWebSocket
 
     override fun onMessage(session: WsSession, text: String, last: Boolean) {
-        wsFor(session).triggerMessage(WsMessage(text))
+        websocket.triggerMessage(WsMessage(text))
     }
 
     override fun onMessage(session: WsSession, buffer: BufferData, last: Boolean) {
-        wsFor(session).triggerMessage(WsMessage(buffer.readBytes()))
+        websocket.triggerMessage(WsMessage(buffer.readBytes()))
     }
 
     override fun onClose(session: WsSession, status: Int, reason: String) {
-        wsFor(session).triggerClose(WsStatus(status, reason))
-        sessions.remove(session)
+        websocket.triggerClose(WsStatus(status, reason))
     }
 
     override fun onError(session: WsSession, t: Throwable) {
-        wsFor(session).triggerError(t)
+        websocket.triggerError(t)
     }
 
     override fun onOpen(session: WsSession) {
-        wsFor(session)
-    }
-
-    private fun wsFor(session: WsSession): PushPullAdaptingWebSocket {
-        return sessions.getOrPut(session) {
-            val uriPath = uriPath.get(prologue.get(session) as HttpPrologue) as UriPath
-            val query = rawQuery.get(prologue.get(session) as HttpPrologue)
-            val upgradeRequest = org.http4k.core.Request(GET, uriPath.rawPath() + "?" + query)
-            object : PushPullAdaptingWebSocket() {
-                override fun send(message: WsMessage) {
-                    when (message.mode) {
-                        Binary -> session.send(BufferData.create(message.body.payload.array()), true)
-                        Text -> session.send(message.bodyString(), true)
-                    }
+        websocket = object : PushPullAdaptingWebSocket() {
+            override fun send(message: WsMessage) {
+                when (message.mode) {
+                    Binary -> session.send(BufferData.create(message.body.payload.array()), true)
+                    Text -> session.send(message.bodyString(), true)
                 }
+            }
 
-                override fun close(status: WsStatus) {
-                    session.close(status.code, status.description)
-                }
-            }.apply(ws(upgradeRequest))
-        }
+            override fun close(status: WsStatus) {
+                session.close(status.code, status.description)
+            }
+        }.apply(ws(upgradeRequest))
     }
 
-    private val prologue = WsConnection::class.java.getDeclaredField("prologue").also {
-        it.isAccessible = true
-    }
-    private val uriPath = HttpPrologue::class.java.getDeclaredField("uriPath").also {
-        it.isAccessible = true
-    }
-    private val rawQuery = HttpPrologue::class.java.getDeclaredField("rawQuery").also {
-        it.isAccessible = true
+    override fun onHttpUpgrade(prologue: HttpPrologue, headers: Headers): Optional<Headers> {
+        assert(!complete) { "Cannot be reused for multiple sessions" }
```

Comment:
I could use an `AtomicBoolean` to make this atomic, but given that it's a testing safeguard, I don't think the performance impact is worth it.